core/executor/src/wasm_executor.rs

```diff
@@ -638,74 +636,109 @@ impl WasmExecutor {
 		method: &str,
 		data: &[u8],
 		) -> Result<Vec<u8>> {
-		let module = ::wasmi::Module::from_buffer(code).expect("all modules compiled with rustc are valid wasm code; qed");
-		self.call_in_wasm_module(ext, heap_pages, &module, method, data)
+		let module = ::wasmi::Module::from_buffer(code)
+			.expect("all modules compiled with rustc are valid wasm code; qed");
+		let module = self.prepare_module(ext, heap_pages, &module)
+			.expect("all modules compiled with rustc are valid wasm code; qed");
+		self.call_in_wasm_module(ext, &module, method, data)
 	}
 
 	/// Call a given method in the given wasm-module runtime.
 	pub fn call_in_wasm_module<E: Externalities<Blake2Hasher>>(
 		&self,
 		ext: &mut E,
-		heap_pages: usize,
-		module: &Module,
+		module_instance: &ModuleRef,
 		method: &str,
 		data: &[u8],
 	) -> Result<Vec<u8>> {
-		// start module instantiation. Don't run 'start' function yet.
-		let intermediate_instance = ModuleInstance::new(
-			module,
-			&ImportsBuilder::new()
-				.with_resolver("env", FunctionExecutor::<E>::resolver())
-		)?;
-
 		// extract a reference to a linear memory, optional reference to a table
 		// and then initialize FunctionExecutor.
-		let memory = intermediate_instance
-			.not_started_instance()
+		let memory = module_instance
 			.export_by_name("memory")
 			// TODO: with code coming from the blockchain it isn't strictly been compiled with rustc anymore.
 			// these assumptions are probably not true anymore
 			.expect("all modules compiled with rustc should have an export named 'memory'; qed")
 			.as_memory()
 			.expect("in module generated by rustc export named 'memory' should be a memory; qed")
 			.clone();
-		let table: Option<TableRef> = intermediate_instance
-			.not_started_instance()
+		let table: Option<TableRef> = module_instance
 			.export_by_name("__indirect_function_table")
 			.and_then(|e| e.as_table().cloned());
 
-		let mut fec = FunctionExecutor::new(memory.clone(), heap_pages, table, ext)?;
-
-		// finish instantiation by running 'start' function (if any).
-		let instance = intermediate_instance.run_start(&mut fec)?;
+		let low = memory.lowest_used();
+		let used_mem = memory.used_size();
```

Comment:
Assuming that `used_size` returns the highest written address so far, this will be set to the address of last written byte (shouldn't it be +1?) and then in `Heap` we assume that we can allocate memory starting from this address.

But because we didn't yet run the start function (which could run arbitrary initialization code) this assumption can be violated.

I think this would work for now since rust doesn't use start function at all atm and we only have rust runtimes now. But I'm not sure that this is universal approach, since the behavior of rust can be changed and/or somebody will want to use a language other than rust. 

Of course, we can impose constraints on wasm modules to be compatible with substrate, however, I think this might be not the decision and could backfire even with Rust.

Reply:
> Assuming that `used_size` returns the highest written address so far, this will be set to the address of last written byte (shouldn't it be +1?) and then in `Heap` we assume that we can allocate memory starting from this address.

+1 is not required. The buffer should be schrinked to exactly this size.

> 
> This completely ignores the fact that compiler might reserve some space without actually writing to it (aka BSS).
> 

Not exactly. This relies on wasmi to always initialize all data segments, as it does now. Since this code already relies on some of the wasmi internals I don't think this is much of a stretch. 

The bottom line is simple. If the interpreter can do 200bps as demonstrated here it should provide a convenient interface to be used efficiently. I'm sure compatibility can be achieved without sacrificing performance.

Reply:
re: the +1 remark, wasmi documentation doesn't appear to be consistent:

```
   /// Returns current used memory size in bytes.
    /// This is the highest memory address that had been written to.
    pub fn used_size(&self) -> Bytes {
        Bytes(self.buffer.borrow().len())
    }
```

i would presume that the `buffer.len()` actually does give a *size*, rather than an *index* (i.e. you write to *index* 0 implies your buffer is *size* 1).

so the remark "This is the highest memory address that had been written to" is wrong: it should read "This is **one more than** the highest memory address that had been written to".

@pepyakin could you confirm?

Reply:
That function was added by me and the documentation is indeed incorrect.